src/include/storage/data_table.h

```diff
@@ -114,39 +101,46 @@ class DataTable {
    private:
     friend class DataTable;
 
-    SlotIterator() : is_end_(true) {}
+    SlotIterator() = default;
 
-    SlotIterator(const DataTable *table) : table_(table), block_index_(0), is_end_(false) {  // NOLINT
+    SlotIterator(const DataTable *table) : table_(table), block_index_(0) {  // NOLINT
       end_index_ = table_->blocks_size_;
       TERRIER_ASSERT(end_index_ >= 1, "there should always be at least one block");
-
-      // only 1 empty block
-      if (end_index_ == 1) {
-        common::SharedLatch::ScopedSharedLatch latch(&table_->blocks_latch_);
-        if (table_->blocks_[0]->GetInsertHead() == 0) {
-          end_index_ = 0;
-        }
-      }
-
       UpdateFromNextBlock();
     }
 
     void UpdateFromNextBlock() {
-      TERRIER_ASSERT(table_->blocks_size_ >= 1, "there should always be at least one block");
-      RawBlock *b;
-      {
-        common::SharedLatch::ScopedSharedLatch latch(&table_->blocks_latch_);
-        b = const_cast<DataTable *>(table_)->blocks_[block_index_];
+      TERRIER_ASSERT(end_index_ >= 1, "there should always be at least one block");
+
+      while (true) {
+        if (UNLIKELY(block_index_ == end_index_)) {
+          max_slot_num_ = 0;
+          current_slot_ = InvalidTupleSlot();
+          return;
+        }
+
+        RawBlock *b;
+        {
+          common::SharedLatch::ScopedSharedLatch latch(&table_->blocks_latch_);
+          b = table_->blocks_[block_index_];
+        }
+        slot_num_ = 0;
+        max_slot_num_ = b->GetInsertHead();
+        current_slot_ = {b, slot_num_};
+
+        if (max_slot_num_ == 0) {
+          block_index_++;
+          continue;
+        }
+        return;
```

Comment:
Nit: I would generally prefer writing the conditional so that the return is in if-block for two reasons:  1) mirrors the conditional structure of the loop, and 2) makes the loop clearer by eliding the (now) unnecessary `continue`.

Reply:
done